File: export/annot/export_annot.pl
#!/usr/bin/perl

use strict;
use utf8;
use DBI;
use Encode;
use Config::INI::Reader;

# reading config
my $conf = Config::INI::Reader->read_file($ARGV[0]);
$conf = $conf->{mysql};

# main
my $dbh = DBI->connect('DBI:mysql:'.$conf->{'dbname'}.':'.$conf->{'host'}, $conf->{'user'}, $conf->{'passwd'});
if (!$dbh) {
    die $DBI::errstr;
}
$dbh->do("SET NAMES utf8");
binmode(STDOUT, ':encoding(utf8)');

my $rev = $dbh->prepare("SELECT MAX(rev_id) AS m FROM tf_revisions");
my $books = $dbh->prepare("SELECT * FROM books");
my $tags = $dbh->prepare("SELECT tag_name FROM book_tags WHERE book_id=?");
my $par = $dbh->prepare("SELECT par_id FROM paragraphs WHERE book_id=? ORDER BY pos");
my $sent = $dbh->prepare("SELECT sent_id, source FROM sentences WHERE par_id=? ORDER BY pos");
my $tf = $dbh->prepare("SELECT tf_id, tf_text FROM tokens WHERE sent_id=? ORDER BY pos");
my $tfrev = $dbh->prepare("SELECT rev_id, rev_text FROM tf_revisions WHERE tf_id=? AND is_last=1");
my $r;

$rev->execute();
$r = $rev->fetchrow_hashref();
my $maxrev = $r->{'m'};

my $header = "<?xml version=\"1.0\" encoding=\"utf-8\" standalone=\"yes\"?>\n<annotation version=\"0.11\" revision=\"$maxrev\">\n";
my $footer = "</annotation>";

print $header;

$books->execute();
while ($r = $books->fetchrow_hashref()) {
    print '<text id="'.$r->{'book_id'}.'" parent="'.$r->{'parent_id'}.'" name="'.tidy_xml(decode('utf8', $r->{'book_name'}))."\">\n  <tags>\n";
    $tags->execute($r->{'book_id'});
    while (my $r1 = $tags->fetchrow_hashref()) {
        print "    <tag>".tidy_xml(decode('utf8', $r1->{'tag_name'}))."</tag>\n";
    }
    print "  </tags>\n";
    print "  <paragraphs>\n";
    $par->execute($r->{'book_id'});
    while (my $r1 = $par->fetchrow_hashref()) {
        print_paragraph($r1->{'par_id'}, $ARGV[1] eq 'no_ambig');
    }
    print "  </paragraphs>\n";
    print "</text>\n";
}

print $footer;

# subroutines
sub tidy_xml {
    my $arg = shift;
    $arg =~ s/&/&amp;/g;
    $arg =~ s/"/&quot;/g;
    $arg =~ s/'/&apos;/g;
    $arg =~ s/</&lt;/g;
    $arg =~ s/>/&gt;/g;
    return $arg;
}
sub print_paragraph {
    my $id = shift;
    my $only_unambiguous = shift;
    my $out = '';
    my $should_print = 0;

    $out .= "    <paragraph id=\"$id\">\n";
    $sent->execute($id);
    while (my $r = $sent->fetchrow_hashref()) {
        my $s = get_sentence($r->{'sent_id'}, $r->{'source'});
        if (!$only_unambiguous || !$s->[1]) {
            $should_print = 1;
            $out .= $s->[0];
        }
    }
    $out .= "    </paragraph>\n";
    print $out if $should_print;
}
sub get_sentence {
    my $id = shift;
    my $source = shift;
    my $out_text = '';
    my $has_ambiguity = 0;

    $out_text .= "      <sentence id=\"$id\">\n";
    $out_text .= "        <source>".tidy_xml(decode('utf8', $source))."</source>\n        <tokens>\n";
    $tf->execute($id);
    while (my $r = $tf->fetchrow_hashref()) {
        my $t = get_token($r->{'tf_id'}, $r->{'tf_text'});
        $out_text .= $t->[0];
        if ($t->[1]) {
            $has_ambiguity = 1;
        }
    }
    $out_text .= "        </tokens>\n      </sentence>\n";
    return [$out_text, $has_ambiguity];
}
sub get_token {
    my $id = shift;
    my $text = shift;
    my $out_text = '';
    my $is_ambiguous = 0;

    $out_text .= "          <token id=\"$id\" text=\"".tidy_xml(decode('utf8', $text))."\">";
    $tfrev->execute($id);
    my $r = $tfrev->fetchrow_hashref();
    my ($rev_id, $rev_text) = ($r->{'rev_id'}, $r->{'rev_text'});

    $text =~ s/<tfr/<tfr rev_id="$rev_id"/;

    my $vars = 0;
    while ($rev_text =~ /<v>/g) {
        if (++$vars > 1) {
            $is_ambiguous = 1;
            last;
        }
    }

    $out_text .= decode('utf8', $rev_text);
    $out_text .= "</token>\n";
    return [$out_text, $is_ambiguous];
}

File: export/annot/export_annot.php
<?php

if (php_sapi_name() != 'cli')
    die("This script is for CLI only");

set_include_path(get_include_path().PATH_SEPARATOR.'/corpus');
require_once('lib/header_ajax.php');

define('VERSION', "0.12");
$TAGS = sql_prepare("SELECT tag_name FROM book_tags WHERE book_id=?");
$PARAGRAPHS = sql_prepare("SELECT par_id FROM paragraphs WHERE book_id=? ORDER BY pos");
$SENTENCES = sql_prepare("SELECT sent_id, source FROM sentences WHERE par_id=? ORDER BY pos");
$TOKENS = sql_prepare("
    SELECT tf_id, tf_text, rev_id, rev_text
    FROM tokens
    JOIN tf_revisions
        USING (tf_id)
    WHERE sent_id=? AND is_last=1
    ORDER BY pos
");

main();


function main() {
    $xw = new XMLWriter();
    $xw->openUri('php://output');
    $xw->startDocument("1.0", "utf-8", "yes");
    $xw->startElement("annotation");
    $maxrev = sql_fetchall(sql_query("SELECT MAX(rev_id) AS m FROM tf_revisions"))[0]['m'];
    $xw->writeAttribute("version", VERSION);
    $xw->writeAttribute("revision", $maxrev);
    $xw->setIndent(true);
    $xw->setIndentString("  ");
    $books = sql_fetchall(sql_query("SELECT * FROM books"));
    foreach ($books as $book) {
        add_book($xw, $book);
    }
    $xw->endElement();
    $xw->endDocument();
}

function add_book($xw, $book) {
    $xw->startElement("text");
    $xw->writeAttribute("id", $book['book_id']);
    $xw->writeAttribute("parent", $book['parent_id']);
    $xw->writeAttribute("name", $book['book_name']);

    global $TAGS;
    $xw->startElement("tags");
    sql_execute($TAGS, array($book['book_id']));
    foreach (sql_fetchall($TAGS) as $tag) {
        $xw->writeElement("tag", $tag['tag_name']);
    }
    $xw->endElement();

    global $PARAGRAPHS;
    $xw->startElement("paragraphs");
    sql_execute($PARAGRAPHS, array($book['book_id']));
    foreach (sql_fetchall($PARAGRAPHS) as $paragraph) {
        add_paragraph($xw, $paragraph);
    }
    $xw->endElement();

    $xw->endElement();
}

function add_paragraph($xw, $paragraph) {
    global $SENTENCES;
    $xw->startElement("paragraph");
    $xw->writeAttribute("id", $paragraph['par_id']);
    sql_execute($SENTENCES, array($paragraph['par_id']));
    foreach (sql_fetchall($SENTENCES) as $sentence) {
        add_sentence($xw, $sentence);
    }
    $xw->endElement();
}

function add_sentence($xw, $sentence) {
    global $TOKENS;
    $xw->startElement("sentence");
    $xw->writeAttribute("id", $sentence['sent_id']);
    $xw->writeElement("source", $sentence['source']);
    $xw->startElement("tokens");
    sql_execute($TOKENS, array($sentence['sent_id']));
    foreach (sql_fetchall($TOKENS) as $token) {
        add_token($xw, $token);
    }
    $xw->endElement();
    $xw->endElement();
}

function add_token($xw, $token) {
    $xw->startElement("token");
    $xw->writeAttribute("id", $token['tf_id']);
    $xw->writeAttribute("text", $token['tf_text']);
    $xw->writeRaw(str_replace("<tfr", "<tfr rev_id=\"".$token['rev_id']."\"", $token['rev_text']));
    $xw->endElement();
}


?>
